comp/otelcol/otlp/config.go

```diff
@@ -62,17 +64,63 @@ func readConfigSection(cfg config.Reader, section string) *confmap.Conf {
 	for _, key := range cfg.AllKeysLowercased() {
 		if strings.HasPrefix(key, prefix) && cfg.IsSet(key) {
 			mapKey := strings.ReplaceAll(key[len(prefix):], ".", confmap.KeyDelimiter)
+
+			if v, ok := typeMap[mapKey]; ok {
```

Comment:
The issue I was seeing the other day was unrelated to types. Rather, the `stringMap` variable would be built to look like this:

```
stringMap = {
  "a::b::c": value1,
  "a": {
    "b": {
      "c": value2
    }
  }
}
```

Because of how `confmap.NewFromStringMap` works, it treats keys with `::` and nested keys as essentially the same thing, meaning the object declared above has two identical keys with different values. The proper fix there would be normalizing the map such that this doesn't happen.

I would recommend that the body of this function gets extracted out so it could be tested directly, leaving it looking like this:

```
func readConfigSection(cfg config.Reader, section string, target interface{}) *confmap.Conf {
  // NOTE: target is nil or the struct being considered for type information
  stringMap := buildStringMap(cfg, section, target)
  return confmap.NewFromStringMap(stringMap)
}
```

Then `buildStringMap` can be tested in isolation to ensure no duplicate keys are created by this 2 step algorithm.